Fixes #660 - Revert only changes of selected deployment on rollback

kolloch opened the pull request:
Also introduces an ordering of PathIds.

I considered extracting the revert logic into a separate class DeploymentPlanRevert with the origin, target fields copied. I'd like to hear feedback of the reviewer about that idea.

drexin commented:
@kolloch Could you please rebase this?

drexin commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentPlan.scala:
```diff
@@ -59,13 +59,164 @@ final case class DeploymentPlan(
+    sortedGroupChanges.foldLeft(group) {
+      case (result, groupUpdate) =>
+        groupUpdate match {
+          case (Some(oldGroup), None) => // revert group removal
```
There's lots of code here. Would it make sense to create methods for the individual cases to increase readability?

kolloch replied:
Ok

drexin commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentPlan.scala:
```diff
@@ -59,13 +59,164 @@ final case class DeploymentPlan(
+    val originalById = mapById(originalSet)
+    val targetById = mapById(targetSet)
+
+    val ids = (originalById.keys ++ targetById.keys).toSeq
```
Could we omit the `toSeq` here? I think `iterator` should be sufficient.

kolloch replied:
yes :)

drexin commented:
Aside from the two comments LGTM.

kolloch commented:
I moved the revert logic into its own class now and I like it :)

kolloch commented:
Don't merge!!!

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentPlanReverter.scala:
```diff
@@ -0,0 +1,165 @@
+      case (result, groupUpdate) =>
+        groupUpdate match {
+          case (Some(oldGroup), None) =>
+            result.update(oldGroup.id, revertGroupRemoval(oldGroup), Timestamp.now())
```
This will create different Timestamps for different changes. Please use one Timestamp for the whole change.

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentPlanReverter.scala:
```diff
@@ -0,0 +1,165 @@
+        .filter { case (oldOpt, newOpt) => oldOpt != newOpt }
+    }
+
+    (revertGroupChanges(groupChanges) _).andThen(revertAppChanges(appChanges))
```
Apply AppChanges before GroupChanges.
Please make a short comment, that the order is important.

aquamatthias commented:
Please adjust also the documentation - pretty sure, this is not covered.

aquamatthias commented:
Made some local tests and now they are working. Good job!
:+1:

kolloch commented:
I hope that I performed all necessary changes:

* I corrected the order of reverting the changes and documented it.
* I corrected the removal of nested apps.
* I made sure that the same Timestamp is used for all changes in the revert.
* I adjusted the wording of the REST API documentation slightly.
* I added another test with nested apps.

kolloch commented:
Thanks for the great reviews!

drexin commented:
:100: This is a great enhancement. Thanks a lot!
